File: omf/core/src/manifest.rs
//! omf manifest schema + clamp rules -- the shared contract (F4 / omf-327).
//!
//! This is the authoritative shape of `omf.toml`. The OMX Go dispatcher
//! (omf/dispatch) MUST mirror this struct exactly so that what the compiled
//! floor validates is what the dispatcher routes.
//!
//! Two invariants are encoded here, not just documented:
//!   1. Command templates are argv ARRAYS, never shell strings -- so the
//!      dispatcher never has to re-quote (the `ai-dispatch.zsh:296-302`
//!      eval-quoting footgun). `Vec<String>` makes a shell string
//!      unrepresentable.
//!   2. The manifest may only ever request STRICTER limits than the compiled
//!      floor. `Limits::clamp_to_floor` enforces this: any looser value is
//!      pulled back to the floor and the clamp is logged. Config can tighten
//!      the floor, never loosen it.

use serde::{Deserialize, Serialize};

use crate::guard::{MLX_CACHE_LIMIT_GIB, MLX_MEMORY_LIMIT_GIB, MLX_WIRED_LIMIT_GIB};

/// Top-level `omf.toml` document.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct Manifest {
    /// Schema version. Bumped on breaking layout changes.
    #[serde(default)]
    pub schema_version: u32,
    /// Declared backends. A backend is the unit of pluggability.
    #[serde(default, rename = "backend")]
    pub backends: Vec<Backend>,
}

/// The kind of upstream a backend fronts. Drives discovery + passthrough.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Kind {
    Forge,
    Claude,
    Codex,
    Omc,
    Omx,
    Gemini,
    Copilot,
    Vendor,
    LocalLlm,
}

/// Routing class -- which HOME/account universe a backend is allowed to enter.
///
/// The compiled work!=private invariant (OMX omf-sq8) keys off this: a `Work`
/// backend can never resolve a private home, and vice versa. `None` is for
/// account-agnostic tools (e.g. a local LLM).
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, Serialize, Deserialize)]
#[serde(rename_all = "kebab-case")]
pub enum Routing {
    #[default]
    None,
    Work,
    Private,
}

/// A single backend row.
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct Backend {
    /// Unique backend name (also the `omf <name>` profile selector).
    pub name: String,
    /// What this backend fronts.
    pub kind: Kind,
    /// Routing class. Defaults to `none` (account-agnostic).
    #[serde(default)]
    pub routing: Routing,
    /// One-shot (`-p`) invocation as an argv array. Empty = not supported.
    #[serde(default)]
    pub oneshot: Vec<String>,
    /// Interactive invocation as an argv array. Empty = not supported.
    #[serde(default)]
    pub interactive: Vec<String>,
    /// Per-child env allowlist: only these names are passed through.
    #[serde(default)]
    pub env_allowlist: Vec<String>,
    /// Whether this backend may run in a danger/yolo mode.
    #[serde(default)]
    pub danger_allowed: bool,
    /// Optional memory limits (local-llm backends). Clamped to the floor.
    #[serde(default)]
    pub limits: Option<Limits>,
}

/// MLX / local-LLM memory limits, in GiB. The manifest may only request
/// values <= the compiled floor; looser requests are clamped.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize)]
pub struct Limits {
    /// Wired (pinned) memory cap.
    pub wired_gib: u64,
    /// Allocator ceiling.
    pub memory_gib: u64,
    /// Cache (returnable-to-OS) ceiling.
    pub cache_gib: u64,
}

impl Default for Limits {
    fn default() -> Self {
        Limits {
            wired_gib: MLX_WIRED_LIMIT_GIB,
            memory_gib: MLX_MEMORY_LIMIT_GIB,
            cache_gib: MLX_CACHE_LIMIT_GIB,
        }
    }
}

/// Record of a single clamp event, for `omf doctor` to surface.
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct ClampLog {
    pub field: &'static str,
    pub requested: u64,
    pub clamped_to: u64,
}

impl Limits {
    /// Clamp each field to the compiled floor. A manifest may tighten (request
    /// a smaller value) but never loosen. Returns the list of clamps applied
    /// (empty when the manifest was already within the floor).
    #[must_use]
    pub fn clamp_to_floor(&mut self) -> Vec<ClampLog> {
        let mut logs = Vec::new();
        let mut clamp = |field: &'static str, val: &mut u64, floor: u64| {
            if *val > floor {
                logs.push(ClampLog {
                    field,
                    requested: *val,
                    clamped_to: floor,
                });
                *val = floor;
            }
        };
        clamp("wired_gib", &mut self.wired_gib, MLX_WIRED_LIMIT_GIB);
        clamp("memory_gib", &mut self.memory_gib, MLX_MEMORY_LIMIT_GIB);
        clamp("cache_gib", &mut self.cache_gib, MLX_CACHE_LIMIT_GIB);
        logs
    }
}

/// Parse a manifest from TOML text. Clamping is the caller's responsibility
/// (via [`Manifest::clamp_all`]) so the raw parse stays side-effect free.
pub fn parse(toml_text: &str) -> Result<Manifest, toml::de::Error> {
    toml::from_str(toml_text)
}

impl Manifest {
    /// Clamp every backend's limits to the floor, returning all clamp events
    /// tagged with the owning backend name.
    #[must_use]
    pub fn clamp_all(&mut self) -> Vec<(String, ClampLog)> {
        let mut all = Vec::new();
        for b in &mut self.backends {
            if let Some(limits) = b.limits.as_mut() {
                for log in limits.clamp_to_floor() {
                    all.push((b.name.clone(), log));
                }
            }
        }
        all
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn empty_stub_parses() {
        let m = parse("schema_version = 0\n").expect("stub parses");
        assert_eq!(m.schema_version, 0);
        assert!(m.backends.is_empty());
    }

    #[test]
    fn backend_row_with_argv_arrays_parses() {
        let txt = r#"
schema_version = 0

[[backend]]
name = "forge-private"
kind = "forge"
routing = "private"
interactive = ["forge"]
oneshot = ["forge", "-p"]
env_allowlist = ["HOME", "PATH", "TERM"]
"#;
        let m = parse(txt).expect("parses");
        assert_eq!(m.backends.len(), 1);
        let b = &m.backends[0];
        assert_eq!(b.kind, Kind::Forge);
        assert_eq!(b.routing, Routing::Private);
        assert_eq!(b.interactive, vec!["forge"]);
        assert_eq!(b.oneshot, vec!["forge", "-p"]);
        assert!(!b.danger_allowed);
    }

    #[test]
    fn looser_limits_are_clamped_to_floor() {
        // Manifest asks for MORE than the floor allows -> must be clamped down.
        let mut l = Limits {
            wired_gib: 99,
            memory_gib: 99,
            cache_gib: 99,
        };
        let logs = l.clamp_to_floor();
        assert_eq!(l.wired_gib, MLX_WIRED_LIMIT_GIB);
        assert_eq!(l.memory_gib, MLX_MEMORY_LIMIT_GIB);
        assert_eq!(l.cache_gib, MLX_CACHE_LIMIT_GIB);
        assert_eq!(logs.len(), 3);
    }

    #[test]
    fn stricter_limits_are_preserved() {
        // Manifest asks for LESS than the floor -> kept as-is, no clamp.
        let mut l = Limits {
            wired_gib: 8,
            memory_gib: 10,
            cache_gib: 1,
        };
        let logs = l.clamp_to_floor();
        assert_eq!(l.wired_gib, 8);
        assert_eq!(l.memory_gib, 10);
        assert_eq!(l.cache_gib, 1);
        assert!(logs.is_empty());
    }
}

File: omf/dispatch/internal/manifest/manifest.go
package manifest

import (
	"fmt"
	"os"
	"path/filepath"
	"strconv"
	"strings"
)

const (
	MLXWiredLimitGiB  uint64 = 14
	MLXMemoryLimitGiB uint64 = 18
	MLXCacheLimitGiB  uint64 = 2
)

// DefaultPath resolves the default location of the omf manifest:
// $HOME/forge/omf.toml. It is the single source of truth shared by the
// dispatcher and `omf doctor` so the two can never resolve different
// manifests. Falls back to a bare relative "omf.toml" only when the home
// directory is completely unresolvable.
func DefaultPath() string {
	if home := HomeDir(); home != "" {
		return filepath.Join(home, "forge", "omf.toml")
	}
	return "omf.toml"
}

// HomeDir returns the user's home directory, preferring os.UserHomeDir() and
// falling back to $HOME. Returns "" only when both are unavailable.
func HomeDir() string {
	if h, err := os.UserHomeDir(); err == nil && h != "" {
		return h
	}
	return os.Getenv("HOME")
}

type Kind string

const (
	KindForge    Kind = "forge"
	KindClaude   Kind = "claude"
	KindCodex    Kind = "codex"
	KindOmc      Kind = "omc"
	KindOmx      Kind = "omx"
	KindGemini   Kind = "gemini"
	KindCopilot  Kind = "copilot"
	KindVendor   Kind = "vendor"
	KindLocalLLM Kind = "local-llm"
)

type Routing string

const (
	RoutingNone    Routing = "none"
	RoutingWork    Routing = "work"
	RoutingPrivate Routing = "private"
)

type Manifest struct {
	SchemaVersion uint32
	Backends      []Backend
}

type Backend struct {
	Name          string
	Kind          Kind
	Routing       Routing
	Oneshot       []string
	Interactive   []string
	EnvAllowlist  []string
	DangerAllowed bool
	Limits        *Limits
}

type Limits struct {
	WiredGiB  uint64
	MemoryGiB uint64
	CacheGiB  uint64
}

type ClampLog struct {
	Backend   string
	Field     string
	Requested uint64
	ClampedTo uint64
}

func Parse(data []byte) (Manifest, error) {
	var m Manifest
	var current *Backend
	inLimits := false

	for lineNo, raw := range strings.Split(string(data), "\n") {
		line := stripComment(raw)
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		switch line {
		case "[[backend]]":
			m.Backends = append(m.Backends, Backend{})
			current = &m.Backends[len(m.Backends)-1]
			inLimits = false
			continue
		case "[backend.limits]":
			if current == nil {
				return Manifest{}, fmt.Errorf("line %d: backend.limits before backend", lineNo+1)
			}
			if current.Limits == nil {
				current.Limits = &Limits{}
			}
			inLimits = true
			continue
		}

		key, val, ok := strings.Cut(line, "=")
		if !ok {
			return Manifest{}, fmt.Errorf("line %d: expected key = value", lineNo+1)
		}
		key = strings.TrimSpace(key)
		val = strings.TrimSpace(val)

		if current == nil {
			if key != "schema_version" {
				return Manifest{}, fmt.Errorf("line %d: top-level key %q is not supported", lineNo+1, key)
			}
			n, err := parseUint(val)
			if err != nil {
				return Manifest{}, fmt.Errorf("line %d: schema_version: %w", lineNo+1, err)
			}
			m.SchemaVersion = uint32(n)
			continue
		}

		if inLimits {
			n, err := parseUint(val)
			if err != nil {
				return Manifest{}, fmt.Errorf("line %d: %s: %w", lineNo+1, key, err)
			}
			switch key {
			case "wired_gib":
				current.Limits.WiredGiB = n
			case "memory_gib":
				current.Limits.MemoryGiB = n
			case "cache_gib":
				current.Limits.CacheGiB = n
			default:
				return Manifest{}, fmt.Errorf("line %d: unsupported limits key %q", lineNo+1, key)
			}
			continue
		}

		switch key {
		case "name":
			s, err := parseString(val)
			if err != nil {
				return Manifest{}, fmt.Errorf("line %d: name: %w", lineNo+1, err)
			}
			current.Name = s
		case "kind":
			s, err := parseString(val)
			if err != nil {
				return Manifest{}, fmt.Errorf("line %d: kind: %w", lineNo+1, err)
			}
			kind := Kind(s)
			if !validKind(kind) {
				return Manifest{}, fmt.Errorf("line %d: unsupported kind %q", lineNo+1, s)
			}
			current.Kind = kind
		case "routing":
			s, err := parseString(val)
			if err != nil {
				return Manifest{}, fmt.Errorf("line %d: routing: %w", lineNo+1, err)
			}
			routing := Routing(s)
			if !validRouting(routing) {
				return Manifest{}, fmt.Errorf("line %d: unsupported routing %q", lineNo+1, s)
			}
			current.Routing = routing
		case "interactive":
			argv, err := parseStringArray(val)
			if err != nil {
				return Manifest{}, fmt.Errorf("line %d: interactive must be an argv array: %w", lineNo+1, err)
			}
			current.Interactive = argv
		case "oneshot":
			argv, err := parseStringArray(val)
			if err != nil {
				return Manifest{}, fmt.Errorf("line %d: oneshot must be an argv array: %w", lineNo+1, err)
			}
			current.Oneshot = argv
		case "env_allowlist":
			argv, err := parseStringArray(val)
			if err != nil {
				return Manifest{}, fmt.Errorf("line %d: env_allowlist must be an array: %w", lineNo+1, err)
			}
			for _, name := range argv {
				if isRoutingManagedEnv(name) {
					return Manifest{}, fmt.Errorf("line %d: env_allowlist %q is routing-managed", lineNo+1, name)
				}
			}
			current.EnvAllowlist = argv
		case "danger_allowed":
			b, err := strconv.ParseBool(val)
			if err != nil {
				return Manifest{}, fmt.Errorf("line %d: danger_allowed: %w", lineNo+1, err)
			}
			current.DangerAllowed = b
		default:
			return Manifest{}, fmt.Errorf("line %d: unsupported backend key %q", lineNo+1, key)
		}
	}

	for i, b := range m.Backends {
		if b.Name == "" {
			return Manifest{}, fmt.Errorf("backend %d: missing name", i)
		}
		if b.Kind == "" {
			return Manifest{}, fmt.Errorf("backend %q: missing kind", b.Name)
		}
	}
	if err := m.Validate(); err != nil {
		return Manifest{}, err
	}
	return m, nil
}

func isRoutingManagedEnv(name string) bool {
	return name == "HOME" || name == "FORGE_CONFIG"
}

func (m Manifest) Validate() error {
	seen := make(map[string]bool)
	for _, b := range m.Backends {
		if seen[b.Name] {
			return fmt.Errorf("backend %q: duplicate name", b.Name)
		}
		seen[b.Name] = true
		if b.Routing == "" {
			return fmt.Errorf("backend %q: missing routing", b.Name)
		}
		if !validRouting(b.Routing) {
			return fmt.Errorf("backend %q: unsupported routing %q", b.Name, b.Routing)
		}
		if b.Routing == RoutingNone && RequiresAccountRouting(b.Kind) {
			return fmt.Errorf("backend %q: kind %q cannot use routing=none", b.Name, b.Kind)
		}
		for _, name := range b.EnvAllowlist {
			if isRoutingManagedEnv(name) {
				return fmt.Errorf("backend %q: env_allowlist %q is routing-managed", b.Name, name)
			}
		}
	}
	return nil
}

func (m *Manifest) ClampAll() []ClampLog {
	var logs []ClampLog
	for i := range m.Backends {
		b := &m.Backends[i]
		if b.Limits == nil {
			continue
		}
		logs = append(logs, clampLimit(b.Name, "wired_gib", &b.Limits.WiredGiB, MLXWiredLimitGiB)...)
		logs = append(logs, clampLimit(b.Name, "memory_gib", &b.Limits.MemoryGiB, MLXMemoryLimitGiB)...)
		logs = append(logs, clampLimit(b.Name, "cache_gib", &b.Limits.CacheGiB, MLXCacheLimitGiB)...)
	}
	return logs
}

func clampLimit(backend, field string, val *uint64, floor uint64) []ClampLog {
	if *val <= floor {
		return nil
	}
	log := ClampLog{Backend: backend, Field: field, Requested: *val, ClampedTo: floor}
	*val = floor
	return []ClampLog{log}
}

func validKind(k Kind) bool {
	switch k {
	case KindForge, KindClaude, KindCodex, KindOmc, KindOmx, KindGemini, KindCopilot, KindVendor, KindLocalLLM:
		return true
	default:
		return false
	}
}

func RequiresAccountRouting(k Kind) bool {
	switch k {
	case KindVendor, KindLocalLLM:
		return false
	default:
		return true
	}
}

func validRouting(r Routing) bool {
	switch r {
	case RoutingNone, RoutingWork, RoutingPrivate:
		return true
	default:
		return false
	}
}

func stripComment(line string) string {
	inString := false
	escaped := false
	for i, r := range line {
		if escaped {
			escaped = false
			continue
		}
		if r == '\\' && inString {
			escaped = true
			continue
		}
		if r == '"' {
			inString = !inString
			continue
		}
		if r == '#' && !inString {
			return line[:i]
		}
	}
	return line
}

func parseString(val string) (string, error) {
	if len(val) < 2 || val[0] != '"' || val[len(val)-1] != '"' {
		return "", fmt.Errorf("expected quoted string")
	}
	return strconv.Unquote(val)
}

func parseUint(val string) (uint64, error) {
	return strconv.ParseUint(strings.TrimSpace(val), 10, 64)
}

func parseStringArray(val string) ([]string, error) {
	val = strings.TrimSpace(val)
	if len(val) < 2 || val[0] != '[' || val[len(val)-1] != ']' {
		return nil, fmt.Errorf("expected []")
	}
	body := strings.TrimSpace(val[1 : len(val)-1])
	if body == "" {
		return nil, nil
	}

	var out []string
	for len(body) > 0 {
		body = strings.TrimSpace(body)
		if !strings.HasPrefix(body, "\"") {
			return nil, fmt.Errorf("expected quoted array element near %q", body)
		}
		end := 1
		escaped := false
		for ; end < len(body); end++ {
			c := body[end]
			if escaped {
				escaped = false
				continue
			}
			if c == '\\' {
				escaped = true
				continue
			}
			if c == '"' {
				break
			}
		}
		if end >= len(body) || body[end] != '"' {
			return nil, fmt.Errorf("unterminated string")
		}
		s, err := strconv.Unquote(body[:end+1])
		if err != nil {
			return nil, err
		}
		out = append(out, s)
		body = strings.TrimSpace(body[end+1:])
		if body == "" {
			break
		}
		if body[0] != ',' {
			return nil, fmt.Errorf("expected comma near %q", body)
		}
		body = body[1:]
	}
	return out, nil
}

func LoadFile(path string) (Manifest, []ClampLog, error) {
	data, err := os.ReadFile(path)
	if err != nil {
		return Manifest{}, nil, err
	}
	m, err := Parse(data)
	if err != nil {
		return Manifest{}, nil, err
	}
	logs := m.ClampAll()
	return m, logs, nil
}
